### ISSUE TITLE
Haskell Server Wrapper Function for Stack specific $PATH

### ISSUE THREAD

**Issue by Pitometsu** [emacs-lsp/lsp-haskell#111]
I have errors like

```
Severity: DsError
Message:  haskell-language-server: could not execute: hspec-discover
```
when run `haskell-language-server`.

But I have no them when run it like `env PATH=$PATH:`bash -c 'source <(stack config env); echo $PATH'` haskell-language-server`.

I can't just run `stack exec haskell-language-server` to just simply have all the `stack`'s environments available, because stack using `nix --pure` for more robust and reproducible builds, and so that cause nix shells conflicts (not sure -- is it because of recursivity, or `stack` can't decide, which nix shell is build one, according to `STACK_IN_NIX_SHELL` environment variable, I guess).

***

So the question is: how construction like `env PATH=$PATH:`bash -c 'source <(stack config env); echo $PATH'` haskell-language-server` could be passed to `lsp-haskell-server-wrapper-function` properly?

**Comment by Pitometsu** [comment]
Would it be something lke:

```elisp
(let
  ((original-path (getenv-internal "PATH"))
   (stack-env-path (shell-command-to-string "stack config env"))
   (args ((mapconcat ’identity argv " "))))
  (list (format "env PATH=%s:%s %s"
   original-path stack-env-path args)))
```
???

**Comment by Pitometsu** [comment]
Also here possible lags with `direnv-mode` and `lorri` tools, and refreshing their caches (like `M-x direnv-update-environment _RET_` and [~~`M-& lorri _RET_`~~](https://github.com/target/lorri/issues/495) `M-x eshell-command _RET_ lorri shell _RET_`

**Comment by Anton-Latukha** [comment]
Maybe the people have difficulty responding, because the case & semantics are not ordered. It is hard to parse the semantics, even weak interpretation does infer the case.

It would be easier to understand if you would describe the environment of the project, what is used for what, and then nail the case.

Overall, from my experience, and from referring to the current official Stack docs, Stack just as 3 years ago, has very weak Nix integration: [link](https://docs.haskellstack.org/en/stable/nix_integration/).

Cabal has much better integration with Nix, running in Nix environment. Cabal would detect that it is running in Nix environment & would use Nix to load dependencies & work as Nix client, download the Haskell binary caches from Nixpkgs Hydra CI, and store the build artifacts in Nix.

Also, important to note that `stack config env` & `stack config --any-key` `stack path` &  `stack path --any-key` are all effectful:
```
$ stack path --project-root

Preparing to install GHC (tinfo6) to an isolated location.
This will not interfere with any system-level installation.
ghc-tinfo6-8.10.4:   10.59 MiB / 207.59 MiB (  5.10%) downloaded...^C
```

Tangent: 
To tell a secret - using `cabal` inside Nix is equivalent to `stack` but with binary caches. As `cabal` would infer the version from Nixpkgs which inherits Stackage every midnight (thou I do not remember Nightly or LTS), and so the Stackage package set is a core of Nixpkgs `haskellPackages`, and all other Hackage set gets weekly handcrafted updates in Nixpkgs to be aligned to whats on Stackage. Thou there is a maintainer change there & also Cabal-Nix setup nevertheless comes with its own configuration woes.

---

But, overall the solution may be as simple as this:
Of course, there is a directive to inherit something from host environment into a pure Nix environment:
```
🅸 ~/s/h/hnix:2021-05-27-refactors◦> echo "$IN_NIX_SHELL"

(nope)
🅸 ~/s/h/hnix:2021-05-27-refactors◦> bash
> export LALALA=TRALALA
> echo "$LALALA"
TRALALA
> nix-shell -p haskellPackages.stack --pure --keep LALALA
[user@host ~/path]$ echo "$IN_NIX_SHELL"
pure
[user@host ~/path]$ echo "$LALALA"
TRALALA
```

Note:
https://github.com/NixOS/nix/issues/419#issuecomment-66932799

:arrow_up: It is true. It was true when it was written, & it is the same currently. Would spare the personal details, but can recommend not do shell work for Nix, or if do any - upfront be ready to throw the work out the window. Nix taught me to not care to contribute when doing contribution, as caring does not depend on me but on one holding the "merge" button.

Also important to note that some "deep" Nix people (seems like principally) - use all bashisms possible, which heavily hardcodes Nix to Bash. So in code implementation, the `/usr/bin/env sh`, preferably `bash` or `dash` need to be run as a wrapper to make `nix-shell -p .. --pure --keep PATH` inheritance work.

Sadly there is a lot of such open secret undocumented technical difficulties which when researched are caused by social context reasons, seems like Conway's law strongly formed many of Nix details in the past.

**Comment by Anton-Latukha** [comment]
Yuriy @Pitometsu, can the topic be closed?

Having somewhat extensive knowledge on the Haskell tooling & Nix situation both from prior work experience ("having skin in the game" with Stack & Nix situation in the past at work, where I build infrastructure pipeline) & my current experience.

The only thing I can tell - is that Stack and Nix combination completely disintegrates in the main agenda of them (the dependency/package management, Stack goes its own way building artifacts while Nix infra demands 100% declarative descriptiveness & reproducibility & integration from Stack, which Stack does not provide for Nix, which produces a result that accumulates downsides of both - not reproducible builds that are not declarative, caching does not work, no deduplication, therefore the set size of the build artifacts during work is tremendous, maintenance is unreasonable, undocumented use, no support from upstreams...). So far Nix & Stack are not designed for each other. Use of Cabal(or HPack or DHall) with Nix is much more happy path & reasonable choice that should succeed.
